src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```diff
@@ -0,0 +1,748 @@
+            case AF_INET6:
+            {
+                constexpr auto sockaddr_in6_size = sizeof(std::remove_pointer_t<LPSOCKADDR_IN6>);
+                if (ptr->ai_addrlen >= sockaddr_in6_size)
+                {
+                    const auto sockaddr_ipv6 = reinterpret_cast<LPSOCKADDR_IN6>(ptr->ai_addr);
+                    char addr[INET6_ADDRSTRLEN] = {};
+                    inet_ntop(AF_INET6, &(sockaddr_ipv6->sin6_addr), addr, sizeof(addr));
+                    ipv6.push_back(addr);
+                    break;
+                }
+                mpl::error(log_category,
+                           "resolve_ip_addresses() -> anomaly: received {} bytes of IPv6 address "
+                           "data while expecting {}!",
+                           ptr->ai_addrlen,
+                           sockaddr_in6_size);
+            }
+            break;
+            default:
+                continue;
+            }
+        }
+    }
+
+    mpl::trace(log_category,
+               "resolve_ip_addresses() -> hostname: {} resolved to : (v4: {}, v6: {})",
+               hostname,
+               fmt::join(ipv4, ","),
+               fmt::join(ipv6, ","));
+
+    return std::make_pair(ipv4, ipv6);
+}
+} // namespace
+
+namespace multipass::hyperv
+{
+
+HCSVirtualMachine::HCSVirtualMachine(const std::string& network_guid,
+                                     const VirtualMachineDescription& desc,
+                                     class VMStatusMonitor& monitor,
+                                     const SSHKeyProvider& key_provider,
+                                     const Path& instance_dir)
+    : BaseVirtualMachine{desc.vm_name, key_provider, instance_dir},
+      description(desc),
+      primary_network_guid(network_guid),
+      monitor(monitor)
+{
+    const auto created_from_scratch = maybe_create_compute_system();
+    const auto state = fetch_state_from_api();
+
+    mpl::debug(log_category,
+               "HCSVirtualMachine::HCSVirtualMachine() > `{}`, created_from_scratch: {}, state: {}",
+               get_name(),
+               created_from_scratch,
+               state);
+
+    // Reflect compute system's state
+    set_state(state);
+    handle_state_update();
+}
+
+void HCSVirtualMachine::compute_system_event_callback(void* event, void* context)
+{
+    mpl::debug(log_category,
+               "compute_system_event_callback() >  event: {}, context: {}",
+               fmt::ptr(event),
+               fmt::ptr(context));
+    const auto type = hcs::parse_event(static_cast<HCS_EVENT*>(event));
+    auto vm = reinterpret_cast<HCSVirtualMachine*>(context);
+
+    switch (type)
+    {
+    case hcs::HcsEventType::SystemExited:
+    {
+        mpl::info(log_category,
+                  "compute_system_event_callback() > {}:  SystemExited event received",
+                  vm->get_name());
+        vm->state = State::off;
+        vm->handle_state_update();
+    }
+    break;
+    case hcs::HcsEventType::Unknown:
```

Comment:
Might add default too, in case someone ever grows the enum grows without keeping this in sync:

```suggestion
    case hcs::HcsEventType::Unknown:
    default:
```